influxdb3_write/src/last_cache.rs

```diff
@@ -0,0 +1,498 @@
+use std::{any::Any, collections::VecDeque, sync::Arc};
+
+use arrow::{
+    array::{
+        ArrayRef, BooleanBuilder, Float64Builder, GenericByteDictionaryBuilder, Int64Builder,
+        RecordBatch, StringBuilder, StringDictionaryBuilder, TimestampNanosecondBuilder,
+        UInt64Builder,
+    },
+    datatypes::{DataType, GenericStringType, Int32Type, SchemaRef, TimeUnit},
+    error::ArrowError,
+};
+use async_trait::async_trait;
+use datafusion::{
+    datasource::{TableProvider, TableType},
+    execution::context::SessionState,
+    logical_expr::{Expr, TableProviderFilterPushDown},
+    physical_plan::{memory::MemoryExec, ExecutionPlan},
+};
+use hashbrown::{HashMap, HashSet};
+use indexmap::IndexMap;
+use iox_time::Time;
+use parking_lot::RwLock;
+use schema::TIME_COLUMN_NAME;
+
+use crate::{
+    catalog::LastCacheSize,
+    write_buffer::{buffer_segment::WriteBatch, FieldData, Row},
+};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("invalid cache size")]
+    InvalidCacheSize,
+    #[error("last cache already exists for database and table")]
+    CacheAlreadyExists,
+}
+
+/// A two level hashmap storing Database Name -> Table Name -> LastCache
+///
+/// There are two lock levels, one at the top and one at the bottom:
+/// - Top: lock the entire cache for creating new entries
+/// - Bottom: lock an individual cache for pushing in new data
+type CacheMap = RwLock<HashMap<String, HashMap<String, RwLock<LastCache>>>>;
+
+pub struct LastCacheProvider {
+    cache_map: CacheMap,
+}
+
+impl std::fmt::Debug for LastCacheProvider {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "LastCacheProvider")
+    }
+}
+
+impl LastCacheProvider {
+    /// Create a new [`LastCacheProvider`]
+    pub(crate) fn new() -> Self {
+        Self {
+            cache_map: Default::default(),
+        }
+    }
+
+    /// Output the records for a given cache as arrow [`RecordBatch`]es
+    #[cfg(test)]
+    pub(crate) fn get_cache_record_batches<D, T>(
+        &self,
+        db_name: D,
+        tbl_name: T,
+    ) -> Option<Result<RecordBatch, ArrowError>>
+    where
+        D: AsRef<str>,
```

Comment:
I'm curious why do the types like this?

Reply:
Sort of a habit I think, but its for a more flexible API. This allows `db_name` or `tbl_name` to be passed as `String`, `&String`, `&str`, `Arc<str>`. I could just make them `&str`, but I find this a bit nicer from the caller side.

For a `#[cfg(test)]` function though... maybe overkill.

Reply:
I've simplified the types in this and some other cases.